src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetInfoService.cs

```diff
@@ -44,25 +45,42 @@ internal abstract class AbstractSnippetInfoService : ForegroundThreadAffinitized
 
         private readonly IAsynchronousOperationListener _waiter;
 
+        /// <summary>
+        /// Collection of tasks started to populate the snippets cache.
+        /// This ensures we wait for them during shutdown.
+        /// </summary>
+        private readonly JoinableTaskCollection _asyncTasks;
+
         public AbstractSnippetInfoService(
             IThreadingContext threadingContext,
             Shell.SVsServiceProvider serviceProvider,
             Guid languageGuidForSnippets,
             IAsynchronousOperationListenerProvider listenerProvider)
             : base(threadingContext)
         {
-            AssertIsForeground();
+            _asyncTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
+            _waiter = listenerProvider.GetListener(FeatureAttribute.Snippets);
+            _languageGuidForSnippets = languageGuidForSnippets;
 
             if (serviceProvider != null)
             {
-                var textManager = (IVsTextManager2)serviceProvider.GetService(typeof(SVsTextManager));
-                if (textManager.GetExpansionManager(out _expansionManager) == VSConstants.S_OK)
+                _asyncTasks.Add(threadingContext.JoinableTaskFactory.RunAsync(() =>
                 {
-                    ComEventSink.Advise<IVsExpansionEvents>(_expansionManager, this);
-                    _waiter = listenerProvider.GetListener(FeatureAttribute.Snippets);
-                    _languageGuidForSnippets = languageGuidForSnippets;
-                    PopulateSnippetCaches();
-                }
+                    return InitializeAndPopulateSnippetsCacheAsync(threadingContext, serviceProvider);
+                }));
+            }
+        }
+
+        private async Task InitializeAndPopulateSnippetsCacheAsync(IThreadingContext threadingContext, Shell.SVsServiceProvider serviceProvider)
+        {
+            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
+            var asyncProvider = (Shell.IAsyncServiceProvider)serviceProvider;
+
+            var textManager = (IVsTextManager2)await asyncProvider.GetServiceAsync(typeof(SVsTextManager)).ConfigureAwait(true);
+            if (textManager.GetExpansionManager(out _expansionManager) == VSConstants.S_OK)
+            {
+                ComEventSink.Advise<IVsExpansionEvents>(_expansionManager, this);
+                await PopulateSnippetCacheAsync().ConfigureAwait(false);
```

Comment:
Is there any concern that more than one PopulateSnippetCacheAsync() could be running at the same time?

Reply:
It was always the case that multiple could be running at the same time (before this change as well) since it just fired off a task.  Since we haven't heard of any issues, and it's running on the UI thread to update the cache we should be safe (there's also locks to update the cache).